File: server/src/routes/reviews.js
const express = require('express');
const prisma = require('../db');
const { authenticateUser } = require('./auth');
const { generateCodeReviewDraft } = require('../services/gemini');
const router = express.Router();

router.use(authenticateUser);

// GET /api/solutions/:id/reviews
router.get('/solutions/:id/reviews', async (req, res) => {
  try {
    const rounds = await prisma.reviewRound.findMany({
      where: { solutionId: req.params.id },
      orderBy: { roundNumber: 'asc' },
      include: {
        reviewer: { select: { id: true, username: true, role: true } }
      }
    });
    res.json({ rounds });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// POST /api/solutions/:id/review/draft (Generate AI Review Draft via Gemini)
router.post('/solutions/:id/review/draft', async (req, res) => {
  if (req.user.role !== 'ADMIN') {
    return res.status(403).json({ error: 'Only admins can generate code review drafts.' });
  }

  try {
    const solution = await prisma.solution.findUnique({
      where: { id: req.params.id },
      include: {
        reviewRounds: { orderBy: { roundNumber: 'asc' } }
      }
    });

    if (!solution) return res.status(404).json({ error: 'Solution not found' });

    const draftText = await generateCodeReviewDraft({
      challengeTitle: solution.challengeTitle,
      language: solution.language,
      code: solution.code,
      previousReviews: solution.reviewRounds
    });

    let parsedDraft = null;
    try {
      parsedDraft = typeof draftText === 'string' ? JSON.parse(draftText) : draftText;
    } catch (e) {
      console.warn("Could not parse Gemini draft JSON:", e.message);
    }

    res.json({
      success: true,
      nextRoundNumber: solution.reviewRounds.length + 1,
      draft: draftText,
      parsedDraft: parsedDraft || { summary: draftText, lineComments: [] }
    });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// POST /api/solutions/:id/review/publish (Publish / Save Review Round)
router.post('/solutions/:id/review/publish', async (req, res) => {
  if (req.user.role !== 'ADMIN') {
    return res.status(403).json({ error: 'Only admins can publish code review rounds.' });
  }

  const { roundNumber, status = 'APPROVED', adminNotes, geminiDraft } = req.body;
  const solutionId = req.params.id;

  if (!adminNotes || !adminNotes.trim()) {
    return res.status(400).json({ error: 'Admin notes/feedback are required to publish a review round.' });
  }

  try {
    const solution = await prisma.solution.findUnique({
      where: { id: solutionId },
      select: { id: true, userId: true, challengeTitle: true }
    });

    if (!solution) {
      return res.status(404).json({ error: 'Solution not found.' });
    }

    const rNum = parseInt(roundNumber) || 1;

    const reviewRound = await prisma.reviewRound.upsert({
      where: {
        solutionId_roundNumber: { solutionId, roundNumber: rNum }
      },
      update: {
        status,
        adminNotes: adminNotes.trim(),
        geminiDraft: geminiDraft ? JSON.stringify(geminiDraft) : null,
        reviewerId: req.user.id
      },
      create: {
        solutionId,
        roundNumber: rNum,
        reviewerId: req.user.id,
        status,
        adminNotes: adminNotes.trim(),
        geminiDraft: geminiDraft ? JSON.stringify(geminiDraft) : null
      },
      include: {
        reviewer: { select: { id: true, username: true, role: true } }
      }
    });

    // Create notification if reviewer is not the solution owner
    if (solution.userId !== req.user.id) {
      await prisma.notification.create({
        data: {
          userId: solution.userId,
          actorId: req.user.id,
          solutionId,
          type: 'REVIEW',
          message: `@${req.user.username} published Review Round #${rNum} (${status}) for "${solution.challengeTitle}"`
        }
      }).catch(err => console.warn('Could not create review notification:', err.message));
    }

    res.json({ success: true, reviewRound });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: server/handlers/reviews.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
	"gorm.io/gorm"
	"hackerrank-server/db"
	"hackerrank-server/models"
	"hackerrank-server/services"
)

// GetReviewsHandler returns all recorded review rounds for a solution
func GetReviewsHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	solutionID := chi.URLParam(r, "id")

	var rounds []models.ReviewRound
	db.DB.Order("roundNumber asc").Preload("Reviewer").Where("solutionId = ?", solutionID).Find(&rounds)

	json.NewEncoder(w).Encode(map[string]interface{}{"rounds": rounds})
}

// GetReviewPromptHandler generates a comprehensive, context-aware LLM review prompt
// containing problem statement, stripped solution code (without input plumbing), and previous rounds/comments.
func GetReviewPromptHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	solutionID := chi.URLParam(r, "id")

	var sol models.Solution
	if err := db.DB.
		Preload("User").
		Preload("ReviewRounds", func(db *gorm.DB) *gorm.DB {
			return db.Order("roundNumber asc").Preload("Reviewer")
		}).
		Preload("Comments", func(db *gorm.DB) *gorm.DB {
			return db.Order("createdAt asc").Preload("User")
		}).
		Where("id = ?", solutionID).First(&sol).Error; err != nil {
		w.WriteHeader(http.StatusNotFound)
		json.NewEncoder(w).Encode(map[string]string{"error": "Solution not found"})
		return
	}

	problemDetails := services.GetProblemDetails(sol.ChallengeSlug)
	promptText := services.BuildReviewPrompt(sol, problemDetails)
	cleanedCode, startLineOffset := services.StripInputPlumbing(sol.Code, sol.Language)

	origLines := strings.Split(strings.ReplaceAll(sol.Code, "\r\n", "\n"), "\n")

	json.NewEncoder(w).Encode(map[string]interface{}{
		"success":          true,
		"prompt":           promptText,
		"challengeTitle":   sol.ChallengeTitle,
		"challengeSlug":    sol.ChallengeSlug,
		"language":         sol.Language,
		"cleanedCode":      cleanedCode,
		"startLineOffset":  startLineOffset,
		"totalLines":       len(origLines),
		"nextRoundNumber":  len(sol.ReviewRounds) + 1,
		"problemStatement": services.HTMLToPlainText(problemDetails.StatementHTML),
	})
}

// PublishReviewHandler records or updates an official code review round decision
func PublishReviewHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	solutionID := chi.URLParam(r, "id")
	currentUser, _ := GetUserFromContext(r.Context())

	if currentUser.Role != "ADMIN" {
		w.WriteHeader(http.StatusForbidden)
		json.NewEncoder(w).Encode(map[string]string{"error": "Only admins can publish code review rounds."})
		return
	}

	var payload struct {
		RoundNumber interface{} `json:"roundNumber"`
		Status      string      `json:"status"`
		AdminNotes  string      `json:"adminNotes"`
		ReviewDraft interface{} `json:"reviewDraft"`
		GeminiDraft interface{} `json:"geminiDraft"` // Backward compatibility
	}

	if err := json.NewDecoder(r.Body).Decode(&payload); err != nil || len(strings.TrimSpace(payload.AdminNotes)) == 0 {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]string{"error": "Admin notes/feedback are required to publish a review round."})
		return
	}

	var sol models.Solution
	if err := db.DB.Where("id = ?", solutionID).First(&sol).Error; err != nil {
		w.WriteHeader(http.StatusNotFound)
		json.NewEncoder(w).Encode(map[string]string{"error": "Solution not found."})
		return
	}

	roundNum := 1
	if payload.RoundNumber != nil {
		switch v := payload.RoundNumber.(type) {
		case float64:
			roundNum = int(v)
		case int:
			roundNum = v
		}
	}

	status := "APPROVED"
	if payload.Status != "" {
		status = payload.Status
	}

	var draftStr *string
	draftInput := payload.ReviewDraft
	if draftInput == nil {
		draftInput = payload.GeminiDraft
	}

	if draftInput != nil {
		switch v := draftInput.(type) {
		case string:
			draftStr = &v
		default:
			if b, err := json.Marshal(v); err == nil {
				s := string(b)
				draftStr = &s
			}
		}
	}

	var round models.ReviewRound
	err := db.DB.Where("solutionId = ? AND roundNumber = ?", solutionID, roundNum).First(&round).Error
	if err != nil {
		round = models.ReviewRound{
			ID:          uuid.New().String(),
			SolutionID:  solutionID,
			RoundNumber: roundNum,
			ReviewerID:  currentUser.ID,
			Status:      status,
			AdminNotes:  payload.AdminNotes,
			GeminiDraft: draftStr,
		}
		db.DB.Create(&round)
	} else {
		db.DB.Model(&round).Updates(map[string]interface{}{
			"status":      status,
			"adminNotes":  payload.AdminNotes,
			"geminiDraft": draftStr,
			"reviewerId":  currentUser.ID,
		})
		round.Status = status
		round.AdminNotes = payload.AdminNotes
		round.GeminiDraft = draftStr
		round.ReviewerID = currentUser.ID
	}

	db.DB.Preload("Reviewer").Where("id = ?", round.ID).First(&round)

	// Create notification if reviewer is not solution owner
	if sol.UserID != currentUser.ID {
		notif := models.Notification{
			ID:         uuid.New().String(),
			UserID:     sol.UserID,
			ActorID:    &currentUser.ID,
			SolutionID: solutionID,
			Type:       "REVIEW",
			Message:    fmt.Sprintf("@%s published Review Round #%d (%s) for \"%s\"", currentUser.Username, roundNum, status, sol.ChallengeTitle),
		}
		db.DB.Create(&notif)
	}

	json.NewEncoder(w).Encode(map[string]interface{}{
		"success":     true,
		"reviewRound": round,
	})
}
